File: src/offer32-3.PrintFromTopToBottom/print_from_top_to_bottom.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:
    vector<vector<int> > Print(TreeNode* pRoot) {
        vector<vector<int>> res;
        if (pRoot == nullptr) return res;
        stack<TreeNode*> sta1;
        stack<TreeNode*> sta2;
        sta1.push(pRoot);
        int len = 0;
        while (!sta1.empty() || !sta2.empty()){
            vector<int> r1, r2;
            while (!sta1.empty()){
                TreeNode* tmp = sta1.top();
                r1.push_back(tmp->val);
                if (tmp->left)
                    sta2.push(tmp->left);
                if (tmp->right)
                    sta2.push(tmp->right);
                sta1.pop();
            }
            if (r1.size()){
                res.push_back(r1);
            }
            while (!sta2.empty()){
                TreeNode* tmp = sta2.top();
                r2.push_back(tmp->val);
                
                if (tmp->right)
                    sta1.push(tmp->right);
                if (tmp->left)
                    sta1.push(tmp->left);
                sta2.pop();
            }
            if (r2.size())
                res.push_back(r2);
        }
        return res;
    }
    
};

File: src/offer32-3.PrintFromTopToBottom/print_from_top_to_bottom.java
import java.util.ArrayList;
import java.util.*;
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public ArrayList<ArrayList<Integer> > Print(TreeNode pRoot) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        if(pRoot == null){
            return res;
        }
        Deque<TreeNode> stack1 = new LinkedList<>();
        Deque<TreeNode> stack2 = new LinkedList<>();
        stack1.push(pRoot);
        int cur = 0;
        while(!stack1.isEmpty() || !stack2.isEmpty()){
            ArrayList<Integer> arr = new ArrayList<>();
            if(cur == 0){
                while(!stack1.isEmpty()){
                    TreeNode tmp = stack1.pop();
                    arr.add(tmp.val);
                    if(tmp.left != null){
                        stack2.push(tmp.left);
                    }
                    if(tmp.right != null){
                        stack2.push(tmp.right);
                    }
                }
            }else{
                while(!stack2.isEmpty()){
                    TreeNode tmp = stack2.pop();
                    arr.add(tmp.val);
                    if(tmp.right != null){
                        stack1.push(tmp.right);
                    }
                    if(tmp.left != null){
                        stack1.push(tmp.left);
                    }
                }
            }
            res.add(arr);
            cur = 1 - cur;
        }
        return res;
    }
}